File: examples/oo/shapes.py
"""
This module defines several shape classes.
"""


class Shape:

    def __init__(self):
        pass

    @property
    def area(self):
        #print("I'm a SHAPE method, not a CIRCLE method!")
        return self._compute_area()

    def _compute_area(self):
        raise NotImplementedError

class Circle(Shape):

    def __init__(self, radius):
        self.radius = radius

    def _compute_area(self):
        return 3.14159*self.radius**2
    
class Rectangle(Shape):
    
    def __init__(self, width, height):
        self.width, self.height = width, height
        
    def _compute_area(self):
        print("Rectangle area!")
        return self.width * self.height

class Square(Rectangle):
    
    def __init__(self, side):
        Rectangle.__init__(self, side, side)
        
 


if __name__ == '__main__':
    A = Shape()
    C = Circle(2.0)
    R = Rectangle(2.0, 4.0)
    S = Square(2.0)
    

File: examples/compiled/cpp_oo/shapes.cc
#include "shapes.hh"

#include <cmath>
#include <iostream>

using std::cout;
using std::endl;

//----------------------------------------------------------------------------//
// SHAPE
//----------------------------------------------------------------------------//

Shape::Shape()
{
	/* ... */
}

double Shape::area()
{
	return this->m_compute_area();
}

//----------------------------------------------------------------------------//
// CIRCLE
//----------------------------------------------------------------------------//

Circle::Circle(double radius)
  : m_radius(radius)
{
	/* ... */
}

double Circle::m_compute_area()
{
	return M_PI*m_radius*m_radius;
}

//----------------------------------------------------------------------------//
// RECTANGLE
//----------------------------------------------------------------------------//


Rectangle::Rectangle(double width, double height)
  : m_width(width), m_height(height)
{
	/* ... */
}

double Rectangle::m_compute_area()
{
	return m_width * m_height;
}


//----------------------------------------------------------------------------//
// RECTANGLE
//----------------------------------------------------------------------------//


Square::Square(double width)
  : Rectangle(width, width)
{
	/* ... */
}

